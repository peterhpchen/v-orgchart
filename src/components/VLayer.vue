<template>
  <table class="v-layer">
    <tr>
      <td>
        <v-node
          :data="data"
        >
          <template
            slot-scope="_"
          >
            <slot
              :data="_.data"
            />
          </template>
        </v-node>
      </td>
    </tr>
    <template
      v-if="isHaveChild"
    >
      <tr
        class="lines"
      >
        <td>
          <div class="downLine"/>
        </td>
      </tr>
      <tr
        class="lines"
      >
        <td
          v-for="n in data.children.length * 2"
          :key="n"
          :class="line(n)"
        />
      </tr>
      <tr
        class="nodes"
      >
        <td
          v-for="(node, index) in data.children"
          :key="key(node, index)"
        >
          <v-layer
            :data="node"
          >
            <template
              slot-scope="_"
            >
              <slot
                :data="_.data"
              />
            </template>
          </v-layer>
        </td>
      </tr>
    </template>
  </table>
</template>

<script>
import VNode from './VNode';

export default {
  name: 'VLayer',
  components: {
    VNode,
  },
  props: {
    data: {
      type: Object,
      required: true,
    },
  },
  computed: {
    isHaveChild() {
      return this.data.children && this.data.children.length > 0;
    },
  },
  methods: {
    key(node, index) {
      return node.id || index;
    },
    line(number) {
      return {
        rightLine: !this.isEven(number),
        leftLine: this.isEven(number),
      };
    },
    isEven(number){
      return number % 2 === 0;
    }
  },
};
</script>

<style lang="postcss" scoped>

</style>

