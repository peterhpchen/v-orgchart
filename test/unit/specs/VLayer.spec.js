import { shallow } from '@vue/test-utils';
import VLayer from '@/components/VLayer';

describe('VLayer.vue', () => {
  describe('given a data with children property', () => {
    it('have two lines tr in the second row', () => {
      const node = {
        name: 'A',
        children: [
          {
            name: 'B',
          },
        ],
      };

      const wrapper = shallow(VLayer, {
        propsData: {
          data: node,
        },
      });

      expect(wrapper.contains('tr:nth-child(2).lines')).toBeTruthy();
    });

    it('have two lines tr in the third row', () => {
      const node = {
        name: 'A',
        children: [
          {
            name: 'B',
          },
        ],
      };

      const wrapper = shallow(VLayer, {
        propsData: {
          data: node,
        },
      });

      expect(wrapper.contains('tr:nth-child(3).lines')).toBeTruthy();
    });

    it('have a nodes tr in the last row', () => {
      const node = {
        name: 'A',
        children: [
          {
            name: 'B',
          },
        ],
      };

      const wrapper = shallow(VLayer, {
        propsData: {
          data: node,
        },
      });

      expect(wrapper.contains('tr:last-child.nodes')).toBeTruthy();
    });

  });

  describe('given a data without children property', () => {
    it('does not have any nodes tr', () => {
      const node = {
        name: 'A',
      };

      const wrapper = shallow(VLayer, {
        propsData: {
          data: node,
        },
      });

      expect(wrapper.contains('.nodes')).toBeFalsy();
    });
  });
});
