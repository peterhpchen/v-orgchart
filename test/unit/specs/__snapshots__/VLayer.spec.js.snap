// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VLayer.vue given a data with children match snapshot 1`] = `
<table class="v-layer">
    <tr>
        <td>
            <!---->
        </td>
    </tr>
    <tr class="lines">
        <td>
            <div class="downLine"></div>
        </td>
    </tr>
    <tr class="lines">
        <td class="rightLine"></td>
        <td class="topLine leftLine"></td>
        <td class="topLine rightLine"></td>
        <td class="topLine leftLine"></td>
        <td class="topLine rightLine"></td>
        <td class="leftLine"></td>
    </tr>
    <tr class="nodes">
        <td>
            <table class="v-layer">
                <tr>
                    <td>
                        <div class="node"></div>
                    </td>
                </tr>
                <!---->
            </table>
        </td>
        <td>
            <table class="v-layer">
                <tr>
                    <td>
                        <div class="node"></div>
                    </td>
                </tr>
                <!---->
            </table>
        </td>
        <td>
            <table class="v-layer">
                <tr>
                    <td>
                        <div class="node"></div>
                    </td>
                </tr>
                <!---->
            </table>
        </td>
    </tr>
</table>
`;
